### PR TITLE
Add rqt plugin for the launch menu settings

### PR DESCRIPTION
**Pull request by Ishadijcks**
## Introduction

The march_rqt_launch_menu is an rqt plugin that makes it easy to select how you want to run the march iv exoskeleton.

It is build with extensibility in mind and therefore it should be incredibly easy to add more launch parameters to the options

## Features
 - Parse Qt forms and upload these settings to the ros parameter server.
 - Basic sanity checks on which run configuration is selected.
 - Add new launch options without adding more code.
 - Ability to change default settings in the ui.
 
## Implementation
The goal of this launch menu was to make it as extensible as possible without having to add more code.

### Getting started
It is highly advised to install Qt creator as it makes editing the .ui files very easy.
```
sudo apt-get install build-essential
sudo apt-get install qtcreator
sudo apt-get install qt5-default
```
In Qt-creator navigate to Design and press `Ctrl + O` to open the `launch_menu.ui` file.

You can now start editing the .ui file!

### Adding a new QCheckBox
You can add a new `QCheckBox` whenever you want to add a new option to the launch menu. It does not matter to which `QGroupBox` this checkbox belongs or where in the layout it's placed.

 1. Drag a Check Box button from the left banner to the `QGroupBox` where you want it to be located.
 2. Double-click the `QCheckBox` to change the text.
 3. In the property editor in the bottom right, locate the `toolTip` and the `statusTip`. 
 4. Change the `toolTip` to a description that fits your setting.
 5. Change the `statusTip` to the name of the parameter that will be uploaded to the parameter server starting with a `/` (e.g. `/ui/example_ui`).
 
 6. Run the launch menu and select the option. Accept by pressing 'OK' in the bottom right.
 7. Check if the parameter is set correctly in the parameter server (e.g. `rosparam get march/launch_menu/ui/example_ui`).
 
### Adding a new QGroupBox
You can add a new QGroupBox whenever you feel the need to group certain buttons together. It is advised to group settings together that are part of the same subtopic (e.g. `/ui/gazebo` and `/ui/rviz`).
 1. Drag the `QGroupBox` to the layout you want it to be a part of.
 2. Double-click the `QGroupBox` to change the title.
 3. Give the object a sensible name in the property editor.
 4. Add whichever checkboxes and radio buttons you want to the `QGroupBox`
 5. Right-click the new `QGroupBox` and change the layout to `Form Layout`.
 6. Optional: If this group will contain radio buttons, change the `statusTip` to the topic you want the setting to publish to (e.g. `/ui`).
 7. The new boxes and buttons will now be automatically loaded!

### Adding a new QRadioButton
It is important to note that only 1 `QRadioButton` can be active in a single `QGroupBox`. Therefore when you want multiple settings involving radio buttons you have to make multiple groups.

If you want to add an option to an existing group the process is quite simple.
 1. Drag the `QRadioButton` to the group.
 2. Give the object a sensible name in the property editor.
 3. Give it a sensible title, this is the value that will be uploaded to the parameter server with spaces removed and in lowercase.
 4. Double-check that the `QGroupBox` it belongs to has a statustip. This is the topic it will be published on.

### Adding a new run configuration
When a new run configuration arises it is relatively simple to add it to the launch menu.
 1. Navigate to `run_configuration_enum.py` and add your run configuration below the existing (e.g. `testjoint = "testjoint"`)
 2. Add a `QRadioButton` to the `RunConfigurationGroup` in Qt creator.
 3. Give it a name that matches the name of your run configuration. Spaces are stripped and it is case-insensitive (e.g. "Test Joint" and "TEs TJOin t" is fine, "test_joint" or "test-joint" is not.)
 4. Optionally give it a `toolTip` to give information to the user hovering over it.
 5. Add a new `QGroupBox` to the `RunConfigurationLayout`.
 6. Give this `QGroupBox` a sensible name in the property editor (e.g. `TestJointGroup`).
 7. Add whatever radio buttons and checkboxes you want.
 8. Right-click the new `QGroupBox` and change the layout to `Form Layout`.
 9. In `launch_menu.py` locate `run_configuration_changed()` and add your group to the if statement.
    ```
    ...
    elif enabled_configuration == run_configuration_enum.testjoint.name:
        self._widget.TestJointGroup.setEnabled(True)
    ...
    ```
10. Run the launch menu to test your new run configuration!
11. Validate it is set correctly with `rosparam get /march/launch_menu/run_configuration`

There are 2 errors that can be thrown when something goes wrong with this section.
 1. You did not add the run configuration to the enumerator, this throws a fatal error.
 2. It was added correctly, but not yet to the if statement. This throws a warning.

### CODE REVIEW

File: march_rqt_launch_menu/plugin.xml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<library path="src">
+    <class name="LaunchMenuPlugin" type="march_rqt_launch_menu.launch_menu.LaunchMenuPlugin" base_class_type="rqt_gui_py::Plugin">
+        <description>
+            An example Python GUI plugin to create a great user interface.
```
Comment (TimBuckers) [inline march_rqt_launch_menu/plugin.xml]:
```suggestion
            GUI with launch settings
```

File: march_rqt_launch_menu/plugin.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+        </description>
+        <qtgui>
+            <label>March Launch Menu</label>
+            <icon>resource/img/march-walking.png</icon>
+            <statustip>Great user interface to provide real value.</statustip>
```
Comment (TimBuckers) [inline march_rqt_launch_menu/plugin.xml]:
```suggestion
            <statustip>Launch Settings</statustip>
```

File: march_rqt_launch_menu/plugin.xml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<library path="src">
+    <class name="LaunchMenuPlugin" type="march_rqt_launch_menu.launch_menu.LaunchMenuPlugin" base_class_type="rqt_gui_py::Plugin">
+        <description>
+            An example Python GUI plugin to create a great user interface.
+        </description>
+        <qtgui>
+            <label>March Launch Menu</label>
+            <icon>resource/img/march-walking.png</icon>
```
Comment (TimBuckers) [inline march_rqt_launch_menu/plugin.xml]: Use different img to distinguish form the input device?

Comment (Ishadijcks) [reply march_rqt_launch_menu/plugin.xml]: Good idea, got anything in mind?

File: march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    def accept_settings(self):
+        dictionary = {}
+
+        # Loop through all groupboxes and get the active radiobutton if they have one.
+        groupboxes = self._widget.findChildren(QGroupBox)
+        for groupbox in groupboxes:
+            if self.has_radio_buttons(groupbox):
+                topic = "/march/launch_settings" + groupbox.statusTip()
+                setting_value = self.get_active_radio_button(groupbox)
```
Comment (TimBuckers) [inline march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py]: This dictionary is not ideal to parse in the state machine launch, we should discuss this before merging

**Comment by TimBuckers** [comment]
@sopje Can you test 'Adding a new run configuration' extensively, I didn't test that part

**Comment by TimBuckers** [comment]
@Ishadijcks do we want to publish 'false' if a radiobutton or checkbox isn't selected? Instead of nothing

File: march_rqt_launch_menu/docs/readme.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ 8. Right-click the new `QGroupBox` and change the layout to `Form Layout`.
+ 9. In `launch_menu.py` locate `run_configuration_changed()` and add your group to the if statement.
+    ```
+    ...
+    elif enabled_configuration == run_configuration_enum.testjoint.name:
+        self._widget.TestJointGroup.setEnabled(True)
+    ...
+    ```
+10. Run the launch menu to test your new run configuration!
+11. Validate it is set correctly with `rosparam get /march/launch_menu/run_configuration`
```
Comment (sopje) [inline march_rqt_launch_menu/docs/readme.md]:
```suggestion
11. Validate it is set correctly with `rosparam get /march/launch_settings/run_configuration`
```

File: march_rqt_launch_menu/docs/readme.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+There are 2 errors that can be thrown when something goes wrong with this section.
+ 1. You did not add the run configuration to the enumerator, this throws a fatal error.
```
Comment (sopje) [inline march_rqt_launch_menu/docs/readme.md]: I have followed all the steps and I indeed get testjoint when I run: rosparam get march/launch_settings/run_configuration.

But now when I want to choose another run configuration, for instance: Simulation --> Gazebo UI --> Smach Viewer, I get the following error:

[INFO] [1551695419.670176]: Uploading the following launch_settings to the parameter server: {'/ui/rqt_status_monitor': False, '/simulation/fake_temperature_sensors': False, '/ui/moveit_rviz': False, '/simulation/gazebo_ui': True, '/ui/rqt_input_device': False, u'/run_configuration': u'simulation', '/ui/smach_viewer': True, u'/testjoint': None}
Traceback (most recent call last):
  File "/home/pm/Documents/march-iv/march_ws/src/state-machine/march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py", line 83, in accept_settings
    self.publish_to_parameter_server(dictionary)
  File "/home/pm/Documents/march-iv/march_ws/src/state-machine/march_rqt_launch_menu/src/march_rqt_launch_menu/launch_menu.py", line 103, in publish_to_parameter_server
    rospy.set_param("march/launch_settings" + key, value)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/client.py", line 514, in set_param
    _param_server[param_name] = param_value #MasterProxy does all the magic for us
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rospy/msproxy.py", line 148, in __setitem__
    self.target.setParam(rospy.names.get_caller_id(), rospy.names.resolve_name(key), val)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1243, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1596, in __request
    allow_none=self.__allow_none)
  File "/usr/lib/python2.7/xmlrpclib.py", line 1094, in dumps
    data = m.dumps(params)
  File "/usr/lib/python2.7/xmlrpclib.py", line 638, in dumps
    dump(v, write)
  File "/usr/lib/python2.7/xmlrpclib.py", line 660, in __dump
    f(self, value, write)
  File "/usr/lib/python2.7/xmlrpclib.py", line 664, in dump_nil
    raise TypeError, "cannot marshal None unless allow_none is enabled"
TypeError: cannot marshal None unless allow_none is enabled
[march_rqt_launch_menu-2] process has died [pid 13257, exit code 1, cmd /home/pm/Documents/march-iv/march_ws/src/state-machine/march_rqt_launch_menu/scripts/rqt_launch_menu __name:=march_rqt_launch_menu __log:=/home/pm/.ros/log/6fb36fea-3e68-11e9-acc8-1c1b0d9a25ca/march_rqt_launch_menu-2.log].
log file: /home/pm/.ros/log/6fb36fea-3e68-11e9-acc8-1c1b0d9a25ca/march_rqt_launch_menu-2*.log

File: march_rqt_launch_menu/scripts/rqt_launch_menu [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+
+import sys
+
+from march_rqt_launch_menu.launch_menu import LaunchMenuPlugin
```
[the remark this answers is not in the document]
Comment (Ishadijcks) [reply march_rqt_launch_menu/scripts/rqt_launch_menu]: Not entirely sure why it's there. Adapted from this guide
http://wiki.ros.org/rqt/Tutorials/Writing%20a%20Python%20Plugin

File: march_rqt_launch_menu/docs/readme.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+You can add a new `QCheckBox` whenever you want to add a new option to the launch menu. It does not matter to which `QGroupBox` this checkbox belongs or where in the layout it's placed.
+
+ 1. Drag a Check Box button from the left banner to the `QGroupBox` where you want it to be located.
+ 2. Double-click the `QCheckBox` to change the text.
+ 3. In the property editor in the bottom right, locate the `toolTip` and the `statusTip`. 
+ 4. Change the `toolTip` to a description that fits your setting.
+ 5. Change the `statusTip` to the name of the parameter that will be uploaded to the parameter server starting with a `/` (e.g. `/ui/example_ui`).
+ 
+ 6. Run the launch menu and select the option. Accept by pressing 'OK' in the bottom right.
+ 7. Check if the parameter is set correctly in the parameter server (e.g. `rosparam get march/launch_menu/ui/example_ui`).
```
[the remark this answers is not in the document]
Comment (Ishadijcks) [reply march_rqt_launch_menu/docs/readme.md]: I agree with the launch_settings, but the example_ui is in line with the example in step 5, so I'd rather leave that as is.

File: march_rqt_launch_menu/scripts/rqt_launch_menu (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ishadijcks) [reply march_rqt_launch_menu/scripts/rqt_launch_menu]: Weirdly enough this import makes it crash for me.

File: march_rqt_launch_menu/docs/readme.md (hunk 6 above)
Comment (Ishadijcks) [reply march_rqt_launch_menu/docs/readme.md]: I have not been able to reproduce this issue, can you send your `.ui` file please?

**Comment by Ishadijcks** [comment]
@sopje I believe to have fixed your issue, can you confirm it works now?

**Comment by Ishadijcks** [comment]
@TimBuckers can this be merged?

**Pull request merged**